### ISSUE TITLE
HasForward type class

### ISSUE THREAD

**Issue by tscholak** [hasktorch/hasktorch#214]
rather than defining the forward function independently from the model data type, we could use a type class that gives users the ability to just call `forward model`. not sure if this is a sound idea, though.

**Comment by tscholak** [comment]
after thinking a bit more about this during my morning commute, I think it could look as simple as this:

```haskell
class HasForward f where
  type a
  forward :: f -> a
```

**Comment by apaszke** [comment]
@tscholak what happens if you use the partial type signatures extension? After all the `Forward f` type should be possible to infer from the `forward` definition. I think we should start using that extension much more heavily, because there are many cases when the type signatures (or more so the constraints!) only make the code less readable.

**tl;dr** Types should highlight the pieces that are important, but everything that's there only for the purpose of having your code be machine-checked for some correctness properties should not pollute your code.

**Comment by tscholak** [comment]
@apaszke the implementation above, https://github.com/hasktorch/hasktorch/issues/214#issuecomment-543201751, was just a first attempt. I didn't try very hard, but what I observed was that it wasn't possible to use partial type signatures and to omit the bulk of the return shape due to the `shape` type variable. for the instance to type-check, it needed to be rigid. I'm sure there's a better way of doing this.

I agree that these insanely complex type signatures should be avoided.

**Comment by tscholak** [comment]
this has been implemented in the meantime
